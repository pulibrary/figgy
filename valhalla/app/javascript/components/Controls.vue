<template>
  <div class="controls">
    <div v-if="orderChanged" id="orderChangedIcon" class="alert alert-info" role="alert">
      <i class="fa fa-exchange"></i> Page order has changed.
    </div>
    <button @click="save" id="save_btn" type="button" class="btn btn-lg btn-primary" :disabled="isDisabled">
      Save
    </button>
  </div>
</template>

<script>
export default {
  name: 'controls',
  computed: {
    id: function () {
      return this.$store.state.id
    },
    resourceClassName: function () {
      return this.$store.state.resourceClassName
    },
    images: function () {
      return this.$store.state.images
    },
    imageIdList: function () {
      return this.$store.getters.imageIdList
    },
    thumbnail: function () {
      return this.$store.state.thumbnail
    },
    startPage: function () {
      return this.$store.state.startPage
    },
    viewingHint: function () {
      return this.$store.state.viewingHint
    },
    viewingDirection: function () {
      return this.$store.state.viewingDirection
    },
    changeList: function () {
      return this.$store.state.changeList
    },
    orderChanged: function () {
      return this.$store.getters.orderChanged
    },
    isDisabled: function () {
      if (this.$store.getters.stateChanged) {
        return false
      } else {
        return true
      }
    },
    fileSetPayload: function () {
      var changed = this.images.filter(image => this.changeList.indexOf(image.id) !== -1 )
      var payload = changed.map((file) => {
        return {id: file.id, title: file.label, page_type: file.page_type }
      })
      return payload
    }
  },
  methods: {
    save: function () {
      let body = {
        resource : {},
        file_sets: this.fileSetPayload
      }
      body.resource[this.resourceClassName] = {
        member_ids: this.imageIdList,
        thumbnail_id: this.thumbnail,
        start_canvas: this.startPage,
        viewing_hint: this.viewingHint,
        viewing_direction: this.viewingDirection,
        id: this.id
      }
      this.$store.dispatch('saveState', body)
    }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style>

</style>
