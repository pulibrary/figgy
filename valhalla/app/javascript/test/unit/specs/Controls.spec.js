import Vuex from 'vuex'
import { mount, createLocalVue } from 'vue-test-utils'
import Controls from '@/components/Controls'
import Fixtures from '@/test/fixtures/image-collection'
const localVue = createLocalVue()
localVue.use(Vuex)

describe('Controls.vue', () => {
  let wrapper
  let options
  let actions
  let getters
  let state
  let store

  beforeEach(() => {
    actions = {
      saveState: jest.fn()
    }
    getters = {
      orderChanged: () => false,
      stateChanged: () => false,
      imageIdList: () => ['e7208ea3-21f3-43d4-9b14-489e15e9791e',
                          '50b5e49b-ade7-4278-8265-4f72081f26a5',
                          'dae7619f-16a7-4306-93e4-70b4b192955c',
                          'b484cd88-fdf2-477c-afe9-d46a49d8822b',
                          '80b02791-4bd9-4566-9a9f-4b3062ba2e0d',
                          '0a3e268f-5872-444e-bdbd-b1a7b01dcb57']
    }
    state = Fixtures.initState
    store = new Vuex.Store({
      getters,
      actions,
      state
    })

    options = {
      computed: {
        orderChanged: function () {
          return getters.orderChanged
        },
        isDisabled: function () {
          if (getters.stateChanged) {
            return false
          } else {
            return true
          }
        }
      }
    }
  })

  it('will not allow save when nothing has changed', () => {
    const wrapper = mount(Controls, { options, store, localVue })
    expect(wrapper.vm.isDisabled).toBeTruthy()
  })

  // make sure saveState button can be tested
  it('allows save once something has changed', () => {
    actions = {
      saveState: jest.fn()
    }
    getters = {
      orderChanged: () => true,
      stateChanged: () => true
    }
    state = Fixtures.initState
    store = new Vuex.Store({
      getters,
      actions,
      state
    })

    const wrapper = mount(Controls, { options, store, localVue })

    expect(wrapper.find('#save_btn').exists()).toBeTruthy()
    wrapper.find('#save_btn').trigger('click')
    expect(actions.saveState).toHaveBeenCalled()
  })

  it('has the expected html structure', () => {
    const wrapper = mount(Controls, { options, store, localVue })
    expect(wrapper.element).toMatchSnapshot()
  })

})
