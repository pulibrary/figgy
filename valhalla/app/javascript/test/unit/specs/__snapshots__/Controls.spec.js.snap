// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls.vue has the expected html structure 1`] = `
<div
  class="controls"
>
  <!---->
   
  <button
    class="btn btn-lg btn-primary"
    disabled="disabled"
    id="save_btn"
    type="button"
  >
    
    Save
  
  </button>
</div>
`;
